python/ray/train/tests/test_ml_trainer.py

```diff
@@ -0,0 +1,68 @@
+import pytest
+
+import ray
+from ray import tune
+
+from ray.ml.preprocessor import Preprocessor
+from ray.train.ml_trainer import Trainer
+from ray.tune.function_runner import wrap_function
+
+
+@pytest.fixture
+def ray_start_4_cpus():
+    address_info = ray.init(num_cpus=4)
+    yield address_info
+    # The code after the yield will run as teardown code.
+    ray.shutdown()
+
+
+class DummyPreprocessor(Preprocessor):
+    def fit_transform(self, ds):
+        return ds.map(lambda x: x * 2)
+
+    def transform(self, ds):
+        return ds.map(lambda x: x + 2)
+
+
+class DummyTrainer(Trainer):
+    def as_trainable(self):
+        def train_func(config):
```

Comment:
Connecting this to what I'm saying above, this is great for a test, but black magic for anyone outside of this company. Compare to:

```python
class Trainer:
    ...

    def training_step(self, *args, **kwargs):
        raise NotImplementedError

class DummyTrainer(Trainer):

    def training_step():
        def train_fn(config):
            tune.report(foo=42)
        return train_fn
```